[OpenWatcom] Can not link with libc compiled in small memory model

Trying to link software using the small-memory model libc gives me:
```
owcc -c -O2 -mcmodel=s -bnone -march=i86 -std=c99 -fno-stack-check -fnostdlib -Wc,-fpi87 -Wc,-zev -Wc,-zls -Wc,-x -Wc,-wcd=303 -Wall -Wextra -I/home/rafael2k/programs/devel/elks/libc/include -I/home/rafael2k/programs/devel/elks/elks/include -I/home/rafael2k/programs/devel/elks/libc/include/watcom -D__ELKS__ -DNDEBUG -o ppmview.obj ppmview.c
owcc -bos2 -s -Wl,option -Wl,dosseg -Wl,option -Wl,start=_start -Wl,option -Wl,nodefaultlibs -Wl,option -Wl,stack=0x1000 -Wl,option -Wl,heapsize=0x1000 -o ppmview ppmview.obj graphics.obj utils.obj -Wl,library -Wl,/home/rafael2k/programs/devel/elks/libc/libc.lib
Error! E2052: file /home/rafael2k/programs/devel/elks/libc/libc.lib(/home/rafael2k/programs/devel/elks/libc/stdio/vfprintf.c): relocation at 0001:14bf not in the same segment
Error! E2052: file /home/rafael2k/programs/devel/elks/libc/libc.lib(/home/rafael2k/programs/devel/elks/libc/stdio/vfprintf.c): relocation at 0001:15aa not in the same segment
Error: Linker returned a bad status
make: *** [Makefile.owc:51: ppmview] Error 1
```

ps: I build the watcom libc after setting watcom.model to "MODEL=s".

--- a/elkscmd/Applications
+++ b/elkscmd/Applications
@@ -209,7 +209,7 @@ test/other/test_float           :test
 #nano-X/bin/nxlandmine           :nanox                   :1232k :1440k
 #nano-X/bin/nxterm               :nanox                   :1232k     :1440c
 #nano-X/bin/nxworld              :nanox                   :1232c :1440k
-nano-X/bin/nxworld.map ::lib/nxworld.map :nanox          :1232c :1440k
+#nano-X/bin/nxworld.map ::lib/nxworld.map :nanox          :1232c :1440k
 basic/basic                     :basic                  :1200k
 advent/advent                     :other                            :1440c
 advent/advent.db ::lib/advent.db  :other                            :1440c

--- a/libc/stdio/vfprintf.c
+++ b/libc/stdio/vfprintf.c
@@ -42,12 +42,25 @@
  * Greg Haerr enhanced for speed, new features
  */
 
+/*
+ * NOTE: Weak symbols aren't working with OpenWatcom small and compact models
+ * due to compiler creating far segment relocation on _weaken() and link failure.
+ * As a result, printf can't support float (%f/%g) or ptick (%k) output
+ * for small and compact model programs without pulling in dtostr and ptostr
+ * whether float/ptick output is used or not. Small and compact model OWC programs
+ * should thus explicitly use dtostr and ptostr for conversion if required.
+ */
+#if defined(__GNUC__) || \
+    (defined(__WATCOMC__) && (defined(__MEDIUM__) || defined(__LARGE__)))
+#define HAS_WEAKEN
+#endif
+
 #include <stdio.h>
 #include <stdlib.h>
 #include <fcntl.h>
 #include <string.h>
 
-#ifndef __HAS_NO_FLOATS__
+#if !defined(__HAS_NO_FLOATS__) && defined(HAS_WEAKEN)
 #include <sys/weaken.h>
 /*
  * Use '#include <sys/weaken.h>` and '__STDIO_PRINT_FLOATS'
@@ -248,7 +261,7 @@ vfprintf(FILE *op, const char *fmt, va_list ap)
          case 'k':              /* Pticks */
           usproc:
             v = lval? va_arg(ap, unsigned long) : (unsigned long)va_arg(ap, unsigned int);
-#ifndef __C86__
+#ifdef HAS_WEAKEN
             if (*fmt == 'k') {
                 if (_weakaddr(ptostr)) {
                     (_weakfn(ptostr))(v, hash, p);
@@ -303,7 +316,7 @@ vfprintf(FILE *op, const char *fmt, va_list ap)
                         buffer_mode);
             break;
 
-#ifndef __HAS_NO_FLOATS__
+#if !defined(__HAS_NO_FLOATS__) && defined(HAS_WEAKEN)
          case 'e':              /* float */
          case 'f':
          case 'g':

